### PR TITLE
Add Chianti collisional data to the new API

### PR DESCRIPTION
**Pull request by epassaro**
:information_source: Continuation of #187 
## Motivation
Adding Chianti collisional data is the last step needed to replicate the atomic files made with past versions of Carsus.

## Changes

- [x] Create `get_lvl_index2id` staticmethod to avoid duplicated code.
- [x] Modify `ChiantiReader` class to grab collisional data and store it in `collisions` DataFrame.
- [x] Write `to_hdf` method for `ChiantiReader` class.
- [x] Modify `TARDISAtomData` to store collisional data and temperatures.
- [x] Changes to `GFALLReader` regarding `to_hdf` method.

## Examples
```
chianti_reader = ChiantiReader(ions='H-He', collisions=True, priority=20)
```

### CODE REVIEW

File: carsus/io/output/base.py [hunk 1]
```diff
@@ -434,6 +438,117 @@ def _create_levels_lines(self, lines_loggf_threshold=-3,
...
+            elif ttype == 5:
+                raise ValueError('Not sure what to do with ttype=5')
+
+            #### 1992A&A...254..436B Equation 20 & 22 #####
+
+            c_ul = 8.63e-6 * upsilon / (g_u * temperatures**.5)
+
+            return pd.Series(data=c_ul, index=c_ul_temperature_cols)
+
+        collisional_strengths = collisions.apply(calculate_collisional_strength, axis=1, args=(temperatures,))
```
Comment (chvogl) [inline carsus/io/output/base.py:546]: This is not a collision strength.
``` 
c_ul = 8.63e-6 * upsilon / (g_u * temperatures**.5)
```
would be the collisional de-excitation rate coefficient (if we multiply it with the electron density); upsilon is the collision strength. We should think about renaming this for clarity.

File: carsus/io/output/base.py [hunk 2]
```diff
@@ -485,6 +600,36 @@ def lines_prepared(self):
...
+    @property
+    def collisions_prepared(self):
+        """
+            Prepare the DataFrame with electron collisions for TARDIS
+            Returns
+            -------
+            collisions_prepared : pandas.DataFrame
+                DataFrame with:
+                    index: atomic_number, ion_number, level_number_lower, level_number_upper;
+                    columns: e_col_id, delta_e, g_ratio, [collision strengths].
```
Comment (chvogl) [inline carsus/io/output/base.py:612]: This is not a collision strength! See previous comment.
We should find a better name for this quantity.

File: carsus/io/chianti_/chianti_.py [hunk 3]
```diff
@@ -602,6 +622,39 @@ def _get_levels_lines(self):
         lines = lines[['energy_upper', 'j_upper', 'energy_lower', 'j_lower',
                        'wavelength', 'gf']]
 
+        if get_collisions:
+            collisions = pd.concat(col_list, sort=True)
+            collisions = collisions.reset_index()
+            collisions = collisions.rename(columns={'lower_level_index': 'level_index_lower',
+                                                    'upper_level_index': 'level_index_upper',
+                                                    'gf_value': 'gf',})
+            # Do we need to fix level indexes here too ?
```
[the remark this answers is not in the document]
Comment (epassaro) [reply carsus/io/chianti_/chianti_.py:631]: Nope. Removing line.

File: carsus/io/output/base.py [hunk 4]
```diff
@@ -434,6 +438,117 @@ def _create_levels_lines(self, lines_loggf_threshold=-3,
...
+        collisions = collisions[['e_col_id', 'lower_level_id',
+                                 'upper_level_id', 'ds_id',
+                                 'btemp', 'bscups', 'ttype', 'cups',
+                                 'gf', 'atomic_number', 'ion_number',
+                                 'level_number_lower', 'g_l',
+                                 'energy_lower', 'level_number_upper', 
+                                 'g_u', 'energy_upper', 'delta_e', 
+                                    'g_ratio']]
+
+        def calculate_collisional_strength(row, temperatures):
```
[the remark this answers is not in the document]
Comment (epassaro) [reply carsus/io/output/base.py:500]: I think not, that's how it was done in the past. I should put that into a staticmethod.

File: carsus/io/output/base.py [hunk 5]
```diff
@@ -434,6 +438,117 @@ def _create_levels_lines(self, lines_loggf_threshold=-3,
...
+                                            upper_levels, on="upper_level_id")
+
+        # Calculate delta_e
+        kb_ev = const.k_B.cgs.to('eV / K').value
+        collisions["delta_e"] = (collisions["energy_upper"] - collisions["energy_lower"])/kb_ev
+
+        # Calculate g_ratio
+        collisions["g_ratio"] = collisions["g_l"] / collisions["g_u"]
+
+        # TODO: pass temperatures as a parameter
```
[the remark this answers is not in the document]
Comment (epassaro) [reply carsus/io/output/base.py]: Done.

File: carsus/io/output/base.py [hunk 6]
```diff
@@ -434,6 +438,117 @@ def _create_levels_lines(self, lines_loggf_threshold=-3,
...
+                x = (kt / delta_e) / (kt / delta_e + c)
+                y_func = interpolate.splev(x, spline_tck)
+                upsilon = y_func / (kt / delta_e + 1)
+
+            elif ttype == 4:
+                x = 1 - np.log(c) / np.log(kt / delta_e + c)
+                y_func = interpolate.splev(x, spline_tck)
+                upsilon = y_func * np.log(kt / delta_e + c)
+
+            elif ttype == 5:
```
[the remark this answers is not in the document]
Comment (epassaro) [reply carsus/io/output/base.py:536]: Not sure. This was written by Misha in `carsus/io/tardis_.py`.

File: carsus/io/output/base.py (hunk 1 above)
Comment (epassaro) [reply carsus/io/output/base.py:545]: I don't remember which name we agreed to use @chvogl.

File: carsus/io/output/base.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (epassaro) [reply carsus/io/output/base.py]: Done.

File: carsus/io/output/base.py (hunk 2 above)
Comment (epassaro) [reply carsus/io/output/base.py]: Done.

**Comment by epassaro** [comment]
> I think this can be merged.
> Do we use astropy.units consistently for all quantities?

What do you mean exactly?

File: carsus/io/output/base.py (hunk 1 above)
Comment (chvogl) [reply carsus/io/output/base.py]: Can you change this to `collisional_ul_factor`(which I think was Wolfgang's preferred option)?

Comment (epassaro) [reply carsus/io/output/base.py]: Done.

File: carsus/io/output/base.py [hunk 7]
```diff
@@ -434,24 +489,87 @@ def _create_levels_lines(self, lines_loggf_threshold=-3,
...
+        # Calculate delta_e
+        kb_ev = const.k_B.cgs.to('eV / K').value
+        collisions["delta_e"] = (collisions["energy_upper"] - collisions["energy_lower"])/kb_ev
+
+        # Calculate g_ratio
+        collisions["g_ratio"] = collisions["g_l"] / collisions["g_u"]
+
+        temperatures = self.collisions_param['temperatures']
+
+        # Derive columns for collisional strenghts
```
Comment (chvogl) [inline carsus/io/output/base.py:535]:
```suggestion
        # Derive columns for collisional strengths
```

File: carsus/io/output/base.py [hunk 8]
```diff
@@ -434,24 +489,87 @@ def _create_levels_lines(self, lines_loggf_threshold=-3,
...
+        collisions = collisions[['e_col_id', 'lower_level_id',
+                                 'upper_level_id', 'ds_id',
+                                 'btemp', 'bscups', 'ttype', 'cups',
+                                 'gf', 'atomic_number', 'ion_number',
+                                 'level_number_lower', 'g_l',
+                                 'energy_lower', 'level_number_upper', 
+                                 'g_u', 'energy_upper', 'delta_e', 
+                                    'g_ratio']]
+
+        collisional_strengths = collisions.apply(self.calculate_collisional_strength,
```
Comment (chvogl) [inline carsus/io/output/base.py:550]:
```suggestion
        collisional_ul_factors = collisions.apply(self.calculate_collisional_strength, 
```

File: carsus/io/output/base.py [hunk 9]
```diff
@@ -434,24 +489,87 @@ def _create_levels_lines(self, lines_loggf_threshold=-3,
...
+                                 'level_number_lower', 'g_l',
+                                 'energy_lower', 'level_number_upper', 
+                                 'g_u', 'energy_upper', 'delta_e', 
+                                    'g_ratio']]
+
+        collisional_strengths = collisions.apply(self.calculate_collisional_strength, 
+                                                 axis=1, args=(temperatures, kb_ev, 
+                                                               c_ul_temperature_cols))
+
+        collisions = collisions.join(collisional_strengths)
```
Comment (chvogl) [inline carsus/io/output/base.py:554]:
```suggestion
        collisions = collisions.join(collisional_ul_factors)
```

Comment (epassaro) [reply carsus/io/output/base.py]: Done.

**Pull request merged**